### ISSUE TITLE
Exposing onColorChangeStart and onColorChangeEnd for Sliders

### ISSUE THREAD

**Issue by osaxma** [rydmike/flex_color_picker#9]
I was wondering if it's possible to expose these two callbacks for the color sliders in a similar manner to the `Slider` Widget. 

An example use case is when the application is keeping track of operations history (undo/redo). With a `Slider` widget, registering every change can be avoided by doing the following:

```dart
Slider(
  onChangeStart: (_) {
    controller.trackChanges = false;
  },
  onChange: (value) {
    controller.updateData(value);
  },
  onChangeEnd: (value) {
    controller.trackChanges = true;
    controller.updateData(value);
  },
)
```

So it'd be nice to have these two callbacks exposed.

**Comment by osaxma** [comment]
By the way, since I'm using the dialog picker, my work around is to stop tracking changes before showing the dialog and resume tracking changes after the dialog is dismissed. So I'm not even tracking changes for anything until a `select` is pressed. 

Though, if the colorPicker isn't shown as a dialog but as a part of the same screen,  it'd be difficult to stop tracking every changes for the slider specifically.

**Comment by rydmike** [comment]
Thanks for the suggestion @osaxma.

Yes I was about to mention too that if you use it in the dialog you can do like the Web example already does, store the start color value, return it if user closed the dialog with cancel or barrier click and then only keep it if OK/SELECT was chosen, and then track the change, **just like you said already in 2nd comment.**

If using it outside a dialog, **yes indeed a solution like you mention would be nice** in order to not track hundreds of changes on the wheel control. The other pickers are just one click, so they don't really start and end, but then again, while there might be tens of them while users play around with color selections, it is not hundreds, like for the wheel (its wheel slider and surface thumb tracker) so for it it would be very useful.

Right now you would have to implement "a use selected color confirmation" button to avoid all that tracking from happening and that is not nice UX.

I will try to get this into the next release I'm working on now. It will be added to null-safe version. Depending on when null-safety goes main stream, I might or might not back port some of the new features I'm adding now to it as well, but leaning more and more towards not adding new features to the none null-safe version, unless they are really critical for someone that won't be able to move the null-safe version yet.

With Web and desktop becoming more mainstream, it is more plausible that the color picker would be used in a none dialog use case, so this would be a good feature to have. It is also useful in full screen dialogs on mobile, where you might just have back option and no cancel.

---
## Off topic

Bunch of new features coming to the null-safe version. Taking the opportunity to add features since it is anyway a new major release. Trying to keep breaking change to a minimum though, other than what null safety entails. Still there will be some minor ones.

### What's in the works
- Major copy paste color value improvements, with keyboard short-cut, top toolbar copy/past buttons, long press copy/paste menu, all optional and configurable based on what is wanted.
- Paste supports parsing multiple RGB color string formats. It automatically detects what is used and auto parses to correct Flutter/dart color value. You can eg paste #RRGGBB RRGGBB #AARRGGBB AARRGGBB #RGB RGB 0xAARRGGBB, partial color values also work.
- For the copy format, the desired resulting RGB color string format can be configured to use #RRGGBB RRGGBB #AARRGGBB AARRGGBB and 0xAARRGGBB options.
- Opacity slider & entry support for color values is coming as well.
- Custom color support on the active picker type selection thumb.
- OK/Cancel as buttons in a dialog top toolbar instead of bottom CANCEL / OK.

Adding your `onChangeStart` `onChangeEnd` event suggestions now as well to the above list.

### Later I'm also planning to add
- Different options for the control that selects used picker. The used `CupertinoSlidingSegmentedControl` offers no keyboard navigation support, probably never will. Going to add a custom version as one option and `ToggleButtons` option as well.
- New advanced color picker, with just sliders, with different color formats HSV, HSL, RGB + maybe some more. The  `onChangeStart` `onChangeEnd` events will apply to these as well.
- More copy color formats for other color standards
- Maybe: A color pipette to pick a color from the screen.

### Other
- The tests and CI/CD for the package needs to get up to same level as FlexColorScheme. Going to work on that as well. 😃

**Comment by osaxma** [comment]
Thank you @rydmike for the detailed response and the information about what's coming up. 

> Opacity slider & entry support for color values is coming as well.
I believe this feature is great to have. 

> OK/Cancel as buttons in a dialog top toolbar instead of bottom CANCEL / OK.
What's the motive for this change? (I actually don't have a preference here, just curious). 

p.s. I appreciate your work and I'm enjoying using the package -- it's simple to use, easy to customize and has excellent documentation. I'll look forward for the coming updates.

**Comment by rydmike** [comment]
Regarding ok / cancel icon buttons in a toolbar, it is just an another option, use it as an alternative or use the current version, can even use both, if so desired.

The onColorChangeStart and onColorChangeEnd are btw working well already in my dev version. I like the feature.

Some new gesture detection improvements in Flutter also enabled me to improve the color wheel gesture handlers so it reacts better than before on taps and drags.

Noticed the ways to improve them while adding the start and end callbacks, so got a double benefit from looking into it.

**Comment by rydmike** [comment]
This feature is now available starting from the pre-release ^2.0.0-nullsafety.1
https://pub.dev/packages/flex_color_picker/versions/2.0.0-nullsafety.2/changelog

You can try and play with the features in the latest web demo as well: https://rydmike.com/flexcolorpicker
The web demo source code is available in the latest package release as well.

These features are still new and still in pre-release status. I expect to migrate 2.0.0 over to stable channel soon after the dust settles from Flutter Engage. 

Let me know if you notice anything that needs further tweaking.

Closing this feature request as it is now implemented. 😃
